flume-ng-core/src/main/java/org/apache/flume/source/ExecSource.java

```diff
@@ -165,30 +165,27 @@
 
   @Override
   public void start() {
-    logger.info("Exec source starting with command:{}", command);
+    logger.info("Exec source starting with command: {}", command);
 
-    executor = Executors.newSingleThreadExecutor();
+    // Start the counter before starting any threads that may access it.
+    sourceCounter.start();
 
-    runner = new ExecRunnable(shell, command, getChannelProcessor(), sourceCounter,
-        restart, restartThrottle, logStderr, bufferCount, batchTimeout, charset);
+    executor = Executors.newSingleThreadExecutor();
+    runner = new ExecRunnable(shell, command, getChannelProcessor(), sourceCounter, restart,
+                              restartThrottle, logStderr, bufferCount, batchTimeout, charset);
 
-    // FIXME: Use a callback-like executor / future to signal us upon failure.
```

Comment:
I'm vaguely guessing that the original comment aimed to take a note about a behavior - which is also specified in the user guide as well - that during startup of ExecSource if the executed command fails it will do that "silently" instead of and this is a TODO.

Example config:
> tier1.sources = source1
> tier1.channels = channel1
> tier1.sinks = sink1
> tier1.channels.channel1.type = file
> tier1.channels.channel1.capacity = 10000
> tier1.sinks.sink1.channel = channel1
> tier1.sources.source1.channels = channel1
> tier1.sources.source1.type = exec
> tier1.sources.source1.command = tail -F /tmp/dirwhichdoesntexist/input.log
> tier1.sinks.sink1.type = logger

Since the user guide also mentions it I'm fine with removing the TODO comment.